proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+
+At this time though there are no practical examples of this pattern. Adding this granularity will add cost to the runtime, as well as to the language as it's hard to reference type parameters in attributes. Going for simplicity at the moment but can reconsider in future versions if evidence for the scenario appears.
+
+### Constraints vs. anti constraints
+The syntax `where T : ref struct` appears in the space C# refers to as constraints. That is the syntax is meant to constraint or limit the set of types that are legal for the generic type parameter.
+
+This feature though expands the set of types available. The set of types applicable to `T where T : ref struct` is larger than simply `T`. This means the feature is an anti-constraint yet it occupies a constraint location. 
+
+It is reasonable to consider a new syntax here that cleanly differentiates our anti-constraints from actual constraints. At the time of this writing though there are only two proposed anti-constraints and one is lacking supporting evidence. This makes it questionable if it meets the bar for a new syntax or not.
+
+If a new syntax is desired to separate anti-constraints from constraints then `allow` is one to consider: 
+
+```c#
+readonly ref struct Span<T> 
+    allow T : ref struct
+{
+
+}
+```
+
+This makes the support for `[SupportsOnlyNonByRefLike]` cleaner as the implementation is to simply ignore the set of `allow` items. It may even lead to a better name such as `[RemoveAllows]`. 
+
+The only downside is that it means type parameter declarations get more complex because there are now both `where` and `allow` lists on them.
+
+```c#
+void Write<T>(T value)
+    where T : ISpanFormattable
+    allow T : ref struct
+{
+
+}
+```
+
+At the same time it's possible this will read clearer to developers. It also opens the door for future allows such as pointers. 
+
+### Generalize the delegate support
+The use of any type in `delegate` may cause readers to desire a more general feature that could be applied to any generic parameter. Essentially could we design a syntax for type parameters that could be any type (pointer, `ref`, etc ...) and allow it on any generic parameter? That would leave `delegate` instances as effectively having an implicit version of this syntax.
+
+It is possible to design such a syntax. Let's use `where T : any` for discussion purposes.
+
+```c#
+// Developer writes
+public delegate T Func<T>();
+
+// Developer effectively writes
+public delegate T Func<T>() where T : any;
+```
+
+The problem with this feature is that it's only useful in signature only scenarios. Those work becase the there is no action associated with the generic parameters. The compiler only needs to consider the standard constraint checking. Effectively the compiler needs to ensure that a `T` with `where T : any` isn't passed to an incompatible generic type parameter.
```

Comment:
```suggestion
The problem with this feature is that it's only useful in signature only scenarios. Those work because there is no action associated with the generic parameters. The compiler only needs to consider the standard constraint checking. Effectively the compiler needs to ensure that a `T` with `where T : any` isn't passed to an incompatible generic type parameter.
```